### ISSUE TITLE
Component props vs. composition

### ISSUE THREAD

**Issue by olegstepura** [react-toolbox/react-toolbox#1191]
There are some attributes in `SnackBar` component  that produce text and a button inside the visible area of `SnackBar` (and it's inside in DOM as well)

Why do we have some attributes like that in `SnackBar` and `AppBar` (maybe more components), not some custom components to be placed as children of those components

So why do we do this:
```js
<Snackbar
  action='Nice'
  label='A new developer started using React Toolbox'
  onClick={this.handleSnackbarClick}
  ref='snackbar'
  type='accept'
/>
```
and not this:
```diff
<Snackbar
-  action='Nice'
-  label='A new developer started using React Toolbox'
-  onClick={this.handleSnackbarClick}
-  type='accept'
  ref='snackbar'
>
+  <Label>A new developer started using React Toolbox</Label>
+  <ActionButton type='accept' onClick={this.handleSnackbarClick}>Nice</ActionButton>
</Snackbar>
```

Is this case it's composable and it's clear that `ProgressBar` must be a child of `SnackBar` as suggested in #1179

**Comment by javivelasco** [comment]
As I mentioned @ #1179 there are some components where it's handy to generate children. An example might be a `Button` where you usually want to place an `Icon` to the left side. Since it's very common, it's just *handy* to provide **optionally** a property to pass an icon that is generated with along `children`.

This doesn't mean it's always the right choice. Take the `Button` again as an example. You can compose your button passing the Icon as a child and it should work too. Consider it as an alternative syntax to get the same behavior/markup.

For the `Snackbar` in particular it's not the best use case. Action buttons should be provided as child components that can be classified to inject new properties, same as the `Layout` works. but the `label` can be useful since it just displays a text properly positioned. It should accept arbitrary children so you are not restricted to the given props though.

In conclusion: those components that generate content from props will be progressively changed to a more composable approach, leaving those props for cases where it's specially handy like the mentioned button. In any case the preferred should be composition.

Let's keep this open so more people can give their opinion!

**Comment by izakfilmalter** [comment]
IMO the Material spec doesn't have the `progress-bar` with `snackbar`. I get the desire to want to put something there, but it doesn't really add any value to the `snackbar`. 

When it comes to composability vs props you run into one main issue in my mind, which is ease of use. It's much easier from a developer pov to pass in a prop. Some components like `Tabs` being a wrapper for `Tab` make a bunch of sense. 

Composability being more functional is a pro, but it does increase the complexity for people using `react-toolbox`. 

Looping back to the Material Spec, it's very opinionated for a reason. There is a strong set of constraints in place that dictate how components should interact and be combined. Yes there is room for creativity in this, but ultimately I think that we should strive to match it to the best of our ability.

**Comment by olegstepura** [comment]
Thanks for a response. 

According `Snackbar` I want to take a note:
* most of it's props are targeted to `Button` actually.
* `Label` in my example can be just text, and thus it will be even better, like this:
  * without button:
  ```js
  <Snackbar>A new developer started using React Toolbox</Snackbar>
  ```
  * with button:
  ```js
  <Snackbar>
     A new developer started using React Toolbox
     <Button type='accept' onClick={this.handleSnackbarClick}>Nice</Button>
  </Snackbar>
  ```

In general case targeting composability is the right choice IMO. Customizing such elements is easier (including, if not the most needed, - CSS styles).

My thoughts are: 
* that's ok to have props that can be replaced with child components (like with `Button`)
* that's not ok to have only ability to provide props. Without ability to use child components that by default take required styling and behavior depending on their context. (e.g. Snackbar passes it's button style overrides to all `Button` children which do not have custom styling passed - not sure if this is doable, though).

**Comment by javivelasco** [comment]
I fully agree! The only purpose of passing props to generate child components is to provide default CSS for those components. Of course you should always be able to give your own. The example you provided with `Snackbar` is perfectly compatible with having some props to generate buttons and a label that set text with some default styles. Every component accepting content should allow to customize that content by passing children, even though it accepts handy props.

I'm keeping this issue open for now so more people can share their thoughts.

**Comment by jgoux** [comment]
A great approach would be to provide both apis, a composable one and composed ones for the really common cases derived from the composable.

What I mean taking the Button as an example : 

```jsx
<Button primary>
  <Icon/>
  I'm a composable button!
</Button>

<IconButton
  primary
  left={<Icon/>}
>
  I'm a composed icon button!
</IconButton>

<PrimaryButton>I'm a composed primary  button!</PrimaryButton>
```

Then it's just a matter of determining what are the most common cases to pick the compisitions, and if they belong to the "core" library or another "extra" library with sane defaults.

**Comment by javivelasco** [comment]
Agree, we can provide the default nesting and the most obvious common cases and then add more from that point
